pkg/cloudprovider/aws/pricing.go

```diff
@@ -0,0 +1,313 @@
+func (p *PricingProvider) InstanceTypes() []string {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	return lo.Union(lo.Keys(p.onDemandPrices), lo.Keys(p.spotPrices))
+}
+
+// OnDemandPrice returns the last known on-demand price for a given instance type, returning an error if there is no
+// known on-demand pricing for the instance type.
+func (p *PricingProvider) OnDemandPrice(instanceType string) (float64, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	price, ok := p.onDemandPrices[instanceType]
+	if !ok {
+		return 0.0, fmt.Errorf("instance type %s not found", instanceType)
+	}
+	return price, nil
+}
+
+// SpotPrice returns the last known spot price for a given instance type, returning an error if there is no
+// known spot pricing for the instance type.
+func (p *PricingProvider) SpotPrice(instanceType string) (float64, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	price, ok := p.spotPrices[instanceType]
+	if !ok {
+		return 0.0, fmt.Errorf("instance type %s not found", instanceType)
+	}
+	return price, nil
+}
+
+func (p *PricingProvider) updatePricing(ctx context.Context) {
+	logging.FromContext(ctx).Infof("updating pricing information")
+	if err := p.updateOnDemandPricing(ctx); err != nil {
+		logging.FromContext(ctx).Errorf("updating on-demand pricing, %s, using existing pricing data from %s", err, p.onDemandUpdateTime.Format(time.RFC822Z))
+	}
+	if err := p.updateSpotPricing(ctx); err != nil {
+		logging.FromContext(ctx).Errorf("updating spot pricing, %s, using existing pricing data from %s", err, p.spotUpdateTime.Format(time.RFC822Z))
+	}
+}
+
+func (p *PricingProvider) updateOnDemandPricing(ctx context.Context) error {
+	prices := map[string]float64{}
+	// standard on-demand instances
+	err := p.fetchOnDemandPricing(ctx, prices,
+		&pricing.Filter{
+			Field: aws.String("tenancy"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("Shared"),
+		},
+		&pricing.Filter{
+			Field: aws.String("productFamily"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("Compute Instance"),
+		})
+	if err != nil {
+		return err
+	}
+
+	// bare metal on-demand prices
+	err = p.fetchOnDemandPricing(ctx, prices,
+		&pricing.Filter{
+			Field: aws.String("tenancy"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("Dedicated"),
+		},
+		&pricing.Filter{
+			Field: aws.String("productFamily"),
+			Type:  aws.String("TERM_MATCH"),
+			Value: aws.String("Compute Instance (bare metal)"),
+		})
+	if err != nil {
+		return err
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.onDemandPrices = prices
+	p.onDemandUpdateTime = time.Now()
+	logging.FromContext(ctx).Infof("updated on-demand pricing with %d instance types", len(p.onDemandPrices))
+	return nil
+}
+
+func (p *PricingProvider) fetchOnDemandPricing(ctx context.Context, prices map[string]float64, additionalFilters ...*pricing.Filter) error {
+	filters := append([]*pricing.Filter{
```

Comment:
consider inlining.

Reply:
It's called twice, one for standard on-demand and once for bare metal.

Reply:
Oh I meant the filters.

Reply:
I prefer to keep function invocations as succinct as possible.